### PR TITLE
Fix add_messages break_up logic

### LINKED ISSUE #281: [More] add_messages > break_up logic is broken
Method [`add_messages`](https://github.com/goavki/phenny/blob/master/modules/more.py#L28) in `modules/more.py` has defaulted parameter `break_up` which dictates how to split the messages to be added.
The default `break_up` logic is given by [`break_up_fn`](https://github.com/goavki/phenny/blob/master/modules/more.py#L10) which enforces the maximum message length of 430.
In part because [this line](https://github.com/goavki/phenny/blob/master/modules/more.py#L30) produces rubbish, most usages first `'\n'.join(x)` and then `x.split('\n')` as a work-around.
If we want to implement #159, we should address this first.

### PR DESCRIPTION
**Pull request by Androbin**
Fixes #281
@shardulc Friendly ping.

### CODE REVIEW

**Comment by Androbin** [comment]
@shardulc Gimme a moment to document the faulty behavior.

**Comment by Androbin** [comment]
* if `msg` is a list, it will split each of them individually.
* if `msg` is a string, it will be wrapped into a singleton list.

The work-around is not needed anymore:
* This simplifies the usage code.
* Messages can be split more thoughtfully.
* Each message will be split at the maximum message length.
* It can be used without much further thought.

**Comment by Androbin** [comment]
This makes the behavior more predictable, therefore preventing bugs like this 032a0285161e5841560023b62c8f1a36020c59e6.

**Comment by Androbin** [comment]
screenshot

**Comment by Androbin** [comment]
@wei2912 If you give me a moment, I'll also provide before-after screenshots with line-breaks.

**Comment by Androbin** [comment]
Just noticed that newlines are stripped out completely.
Isn't there a way to encode newlines over IRC?

**Comment by Androbin** [comment]
screenshot

**Comment by wei2912** [comment]
No, there is no way. The message sent must not contain any new lines, as a
new line would indicate the start of the next command.

**Comment by Androbin** [comment]
Okay, I see.
But instead of `Lorem ipsumdolor sit amet`, how about `Lorem ipsum | dolor sit amet` or something?

**Comment by Androbin** [comment]
* As of your hint, I have discarded the newline splitting.
* I have replaced `input.replace('\n', '')` with `input = re.sub(' ?\n ?', ' ', input)`.
* I took appropriate care of `\r`, byte objects and the maximum message length as well.

**Comment by wei2912** [comment]
Could you show a screenshot of the bot dealing with very long messages?

**Comment by Androbin** [comment]
* I am generating the commit messages using [www.loremipsum.de](http://www.loremipsum.de/), so don't wonder about possibly repeating text.
* The 5 commits got split into 25 messages by `break_up_fn` as expected
screenshot1
screenshot2

**Comment by wei2912** [comment]
That doesn't execute the desired behavior, as ". more 1" should produce strictly one line.

**Comment by Androbin** [comment]
> That doesn't execute the desired behavior, as ". more 1" should produce strictly one line.

Oops, found the problem: `show_more` was using `phenny.reply` which prefixed the already split message and `irc.py` split it again.

**Comment by Androbin** [comment]
This should well account for the occasional lonely lines.

**Comment by Androbin** [comment]
screenshot

**Comment by shardulc** [comment]
@Androbin I believe long commit messages should not be displayed in full at all, like I said before on #291. But the `git.py` module is likely an exception.

**Comment by Androbin** [comment]
> But the git.py module is likely an exception.

Also, messages are truncated by that module.
There is no way any other module could accidentally recreate the truncated part of the message.

**Comment by shardulc** [comment]
They do not appear to be truncated in your screenshot... maybe you tested it before #291 was merged?

**Comment by Androbin** [comment]
> maybe you tested it before #291 was merged?

I beg your pardon, #291 is not yet merged.

**Comment by Androbin** [comment]
* currently, `truncate` only truncates at spaces, not newlines
* currently, `irc.py` splits at spaces and newlines
* currently, `more.py` uses `break_up_fn` which just splits at `max_length`

**Comment by shardulc** [comment]
Right, so how does the approach in my comment sound? `truncate` and `break_up` would be separate functions in `tools.py`.

**Comment by Androbin** [comment]
* I can make `truncate` truncate at newlines as well
* I can refactor the code from `irc.py` into a standalone function `break_up`
* `more.py` should store messages as a whole but let `irc.py` print them using `break_up`

**Comment by shardulc** [comment]
> `more.py` should store messages as a whole but let `irc.py` print them using `break_up`

Why? Why can't `more.py` use `break_up` too? Otherwise `irc.py` would print entire long messages in multiple pieces all at once, which is exactly what `more.py` wants to control.

> I can make truncate truncate at newlines as well

Sure, but that's not really required right now, as it would only be used by `git.py` which doesn't produce newlines. We can make that change later.

**Comment by Androbin** [comment]
* If someone is using `phenny.msg` directly, it splits the message.
* If someone is using `more.add_messages` instead, only part of a split message may be printed
* The more module should give you N messages, not N pieces of different messages.
* Modules which may suffer from long messages may want to truncate them.
* Besides, `irc.py` truncates everything after 3 messages anyway.

**Comment by shardulc** [comment]
I think there's a little misunderstanding. Here's what I'm suggesting:
 - Case 1: Module uses `phenny.msg` directly. The message goes to `irc.py`. Then it is sent to `break_up` in `tools.py` which breaks it up into many pieces, at spaces or newlines. Then `irc.py` prints at most 3 pieces.
 - Case 2: Module uses `more.py`. The messages go to `more.py`. Then they are sent to `break_up` in `tools.py` which breaks them up into many pieces, at spaces or newlines. Then `more.py` prints out an appropriate number of pieces according to commands.
 - Case 3: Module uses `truncate` in `tools.py`. The message is truncated. Then the module uses `phenny.msg` and no breaking happens.

**Comment by Androbin** [comment]
* Case 4: `git.py` uses `tools.truncate` and `more.add_messages` ;)

File: irc.py [hunk 1]
```diff
@@ -229,12 +213,7 @@ def msg(self, recipient, text):
                 self.sending.release()
                 return
 
-        def safe(input): 
-            if type(input) == str:
-                input = input.encode('utf-8')
-            input = input.replace(b'\n', b'')
-            return input.replace(b'\r', b'')
-        self.__write((b'PRIVMSG', safe(recipient)), safe(text))
+        self.write((b'PRIVMSG', recipient), text)
```
Comment (shardulc) [inline irc.py]: Don't you have to pass `text` through `safe` that you have changed above?

Comment (Androbin) [reply irc.py]: `write` is the `safe` version of `__write`

File: modules/more.py [hunk 2]
```diff
@@ -4,42 +4,31 @@
...
-    msgs = break_up(str(msg), max_length)
-    caseless_nick = target.casefold()
+def add_messages(target, phenny, msgs):
+    if not type(msgs) is list:
+        msgs = [msgs]
 
     if not target in phenny.config.channels:
         msgs = list(map(lambda msg: target + ': ' + msg, msgs))
 
+    msgs = sum(map(lambda msg: break_up(msg), msgs), [])
```
Comment (shardulc) [inline modules/more.py]: `sum`? What does this do?

Comment (Androbin) [reply modules/more.py]: `sum` iterates over `map` and adds elements to `[]`

Comment (shardulc) [reply modules/more.py]: You could just use `list` like the line above it.

**Comment by shardulc** [comment]
@Androbin OK, this is looking good in theory now. Can you provide screenshots? (Only for case 1 and case 2.)

File: modules/more.py (hunk 2 above)
Comment (Androbin) [reply modules/more.py]: `list` produces `[['lorem'], ['ipsum', 'dolor'], ['sit', 'amet', 'consetetur']]`
`sum` produces `['lorem', 'ipsum', 'dolor', 'sit', 'amet', 'consetetur']`

**Comment by Androbin** [comment]
Turn out the original `irc.py` splitting behavior is buggy. I need to clean this up first.

**Comment by Androbin** [comment]
screenshot1
screenshot2

File: tools.py [hunk 3]
```diff
@@ -13,12 +13,86 @@
...
+    global MAX_MSG_LEN
+
+    if hasattr(phenny.config, 'MAX_MSG_LEN'):
+        MAX_MSG_LEN = phenny.config.MAX_MSG_LEN
+
+def break_up(text, max_length=MAX_MSG_LEN, max_count=None):
+    if isinstance(text, str): 
+        try:
+            text = text.encode('utf-8')
+        except UnicodeEncodeError:
```
Comment (shardulc) [inline tools.py]: Would prefer the error to be uncaught so that we can see where it came from.

Comment (shardulc) [reply tools.py]: In fact, what is the need for UTF-8 encoding anyway? In Python 3 everything is Unicode.

Comment (Androbin) [reply tools.py]: Okay, I'm open ears. What about these two lines?
https://github.com/goavki/phenny/blob/master/irc.py#L178
https://github.com/goavki/phenny/blob/master/irc.py#L182

Comment (Androbin) [reply tools.py]: The encoding/decoding translates from `str` to `bytes` and back.
Some characters use more than one byte and the limit is on bytes not characters.

Comment (Androbin) [reply tools.py]: What do you mean by "reported" and "in full" ?
"reported" via the `text`, using `print` ?
"in full" as in "the full stack trace" ?
You didn't even let me print it in the git server.

Comment (shardulc) [reply tools.py]: Haha, I believe something like what you did [here](https://github.com/goavki/phenny/blob/master/modules/git.py#L111) will work. Thanks for pointing out the bytes vs. chars.

**Comment by shardulc** [comment]
Tests are passing and the screenshots show the expected behavior (albeit from the wrong module :P but that is not a concern due to #291). Merging this.

**Pull request merged**